### DISCUSSION TITLE
JSON-API: No Authorization with valid token sent

### DISCUSSION THREAD

**Discussion by Thinner77** [awawa-dev/HyperHDR#543] [General]
#### Steps to reproduce
API Authentication: on
Internet API Access: off
Local Admin API Authentication: on
Local API Authentication: off

curl http://localhost:8090/json-rpc -H 'Content-Type: application/json' -H 'Authorization : token [valid token]' -d '{"command":"config","subcommand":"getconfig"}'

#### What is expected?
output of current config

#### What is actually happening?
result of curl request:
{
    "command": "config",
    "error": "No Authorization",
    "success": false,
    "tan": 0
}

When Local Admin API Authentication is disabled, request will succeed.

Maybe my curl request or my understanding of the security system is wrong, but please have a look at this:

JsonAPI::handleConfigCommand() checks for _adminAuthorized
```
void JsonAPI::handleConfigCommand(const QJsonObject& message, const QString& command, int tan)
{
...
else if (subcommand == "getconfig")
	{
		if (_adminAuthorized)
			sendSuccessDataReply(QJsonDocument(_hyperhdr->getQJsonConfig()), full_command, tan);
		else
			sendErrorReply("No Authorization", command, tan);
	}
```
API::isTokenAuthorized() is not setting _adminAuthorized but _authorized:

```
bool API::isTokenAuthorized(const QString& token)
{
	(_authManager->thread() != this->thread())
		? QMetaObject::invokeMethod(_authManager, "isTokenAuthorized", Qt::BlockingQueuedConnection, Q_RETURN_ARG(bool, _authorized), Q_ARG(QString, token))
		: _authorized = _authManager->isTokenAuthorized(token);

	return _authorized;
}
```

#### System

HyperHDR Server: 
- Build:           master (GitHub-bc24df7/a9a00f9-1678986833)
- Build time:      Mar 23 2023 09:35:26
- Git Remote:      https://github.com/awawa-dev/HyperHDR.git
- Version:         20.0.0.0beta0
- UI Lang:         en (BrowserLang: de)
- UI Access:       default
- Avail Capt:      Linux (V4L2)
- Database:        read/write

HyperHDR Server OS: 
- Distribution:   Raspbian GNU/Linux 11 (bullseye)
- Architecture:   arm
- CPU Model:      ARMv6-compatible processor rev 7 (v6l)
- CPU Type:       Raspberry Pi Zero W Rev 1.1
- CPU Revision:   9000c1
- CPU Hardware:   BCM2835
- Kernel:         linux (6.1.19+ (WS: 32))
- Qt Version:     5.15.2
- Browser:        Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:109.0) Gecko/20100101 Firefox/111.0

**Comment by awawa-dev** [comment]
> When Local Admin API Authentication is disabled, request will succeed.

When enabled, you first need to ask the user (a dialog will appear on the HyperHDR page if it is open) to authorize the token before using it (authorize>requestToken). I still can't remember if it can be done with http://ip:8090/json-rpc or json port (default 19444).

**Comment by Thinner77** [comment]
Hi,

token was created directly in Web UI and i can see it in Token Management. There is a requestToken subcommand and according to the docs a popup should appear in the Web UI to accept/reject the token request. But there seems to be no difference in token creation.

Hyperion.ng has the same problem, there is an old (05/2021) issue about this topic, but i don't expect to get that fixed. This is one of the reasons i try to switch to HyperHDR, because i hope you are more motivated and/or skilled and have a better coding style.

A request to
`http://ip:8090/json-rpc?request={%22command%22:%22authorize%22,%22subcommand%22:%22requestToken%22,%22comment%22:%22testtokentest%22}`

results in:
```
{
    "command": "authorize",
    "error": "Command not implemented",
    "success": false,
    "tan": 0
}
```
BTW: All references to _authorized are in API.cpp and i couldn't find any code checking _authorized, it's always set only. This may be caused by the coding style of hyperion.ng developers, maybe they have removed it, i don't know.  

Simple fix of the issue would be to set _adminAuthorized = _authorized; at the end of API::isTokenAuthorized(), but i don't know if this will break something else in security system.

wbr

**Comment by awawa-dev** [comment]
_adminAuthorized is different from _authorized and requires user interaction: just creating a token is not enough. If you don't want the user to interfere, the local admin option should be disabled (but it's enabled on default and the user has to disable it manually). Even if I redesign it to make it work over https POST (http & GET requests are too risky), user interaction will still be required.

> A request to
....
results in:

yes I checked it: it's disabled for http/https requests so only json api RPC port can be used.

**Comment by Thinner77** [comment]
Some more tests:

http://192.168.32.107:8090/json-rpc?request={"command":"config","subcommand":"getconfig"}
{
    "command": "config",
    "error": "No Authorization",
    "success": false,
    "tan": 0
}

http://192.168.32.107:8090/json-rpc?request={"command":"authorize","subcommand":"login","token":"911faca7-0d11-4e06-9424-e46c6c6784b0"}
{
    "command": "authorize",
    "error": "Command not implemented",
    "success": false,
    "tan": 0
}

curl --request POST http://localhost:8090/json-rpc -H 'Content-Type: application/json' --data-raw '{"command":"config","subcommand":"getconfig"}'
{
    "command": "config",
    "error": "No Authorization",
    "success": false,
    "tan": 0
}

curl --request POST http://localhost:8090/json-rpc -H 'Content-Type: application/json' --data-raw '{"command":"authorize","subcommand":"login","token":"911faca7-0d11-4e06-9424-e46c6c6784b0"}'
{
    "command": "authorize",
    "error": "Command not implemented",
    "success": false,
    "tan": 0
}

curl --request POST http://localhost:8090/json-rpc -H 'Content-Type: application/json' -H 'Authorization : token 911faca7-0d11-4e06-9424-e46c6c6784b0' --data-raw '{"command":"config","subcommand":"getconfig"}'
{
    "command": "config",
    "error": "No Authorization",
    "success": false,
    "tan": 0
}

echo '{"command":"config","subcommand":"getconfig"}' | nc localhost 19444
{"command":"config","error":"No Authorization","success":false,"tan":0}
waiting for input... need to CTRL-C

echo '{"command":"authorize","subcommand":"login","token":"911faca7-0d11-4e06-9424-e46c6c6784b0"}' | nc localhost 19444
{"command":"authorize-login","success":true,"tan":0}
waiting for input... need to CTRL-C

Looks like JSON-API can be used by raw access to port 19444 only, but there is no way to use it in shell scripts (would like to switch V4L device input by shell script triggered by lirc irexec, executing getconfig, using jq to change, and executing setconfig to save config).

wbr

**Comment by awawa-dev** [comment]
You can use any websocket client e.g. `websocat` to make it work
```echo '{"command":"authorize","subcommand":"requestToken","comment":"911faca7-0d11-4e06-9424-e46c6c6784b0"}' | ./websocat ws://IP:8090``` will work (ask a user for permission and wont throw "Command not implemented" exception.
If you have lirc and HyperHDR installed on the same machine it can be solved in other way: instead of using JSON API prepare 2 configs, on lirc command stop the HyperHDR service, replace working config with one of two config backup and restart the service.

**Comment by awawa-dev** [comment]
Just change the token if it exists in your config for this example to work: a message in HyperHDR page will popup when you execute the command.

**Comment by Thinner77** [comment]
Some little success:

id has to be set to 5 random chars
websocat -n1 option is needed to wait for response

`echo '{"command":"authorize","subcommand":"requestToken","comment":"testtokentest","id":"12345"}' | websocat -n1 ws://localhost:8090`
-> Popup in WebUI showing
App: testtokentest
Code: 12345

-> JSON result Grant
`{"command":"authorize-requestToken","info":{"comment":"testtokentest","id":"12345","token":"d9f2c817-9b8e-4358-9133-995e611b09ab"},"success":true,"tan":0}`

-> JSON result Deny
`{"command":"authorize-requestToken","error":"Token request timeout or denied","success":false,"tan":0}`

OK, now i have a token d9f2c817-9b8e-4358-9133-995e611b09ab

As I said before creating a token can also be done in WebUI->Advanced->Network Services->Token Management->Create Token
-> d48c072d-3e37-42d2-aad5-5d09162ce1be

It doesn't make any difference in db:
```
sqlite3 -line .hyperhdr/db/hyperhdr.db "select * from auth"

        user =
    password =
       token = 886574156bc31dfe690e7cb89c64841ddf01aa4b71aa2a650498b305d9b2702edaca16c82e7e1484871a7c5310b536b5f42cdaf6773acd0c4dba79cb90f7c4e5
        salt =
     comment = testtokentest
portal_token =
          id = 12345
  created_at = 2023-03-29T18:39:29Z
    last_use =

        user =
    password =
       token = 075035bc903b0c0cb009541f26969d98f4d113a7cee98e7a007bfd8daa9551f00838aef3a693cf14a9d58a2b6300bf70dd97a1535cc65b28cc67a5f817d20e66
        salt =
     comment = createdInWebUI
portal_token =
          id = 15018
  created_at = 2023-03-29T19:29:15Z
    last_use =
```
tokens are saved as SHA512 hashes in db

SHA512(d9f2c817-9b8e-4358-9133-995e611b09ab) = 886574156bc31dfe690e7cb89c64841ddf01aa4b71aa2a650498b305d9b2702edaca16c82e7e1484871a7c5310b536b5f42cdaf6773acd0c4dba79cb90f7c4e5

SHA512(d48c072d-3e37-42d2-aad5-5d09162ce1be) = 075035bc903b0c0cb009541f26969d98f4d113a7cee98e7a007bfd8daa9551f00838aef3a693cf14a9d58a2b6300bf70dd97a1535cc65b28cc67a5f817d20e66


So next task is to request e.g. config:
```
echo '{"command":"config","subcommand":"getconfig"}' | websocat -n1 ws://localhost:8090
{"command":"config","error":"No Authorization","success":false,"tan":0}
```

Lets try with token:
```
echo '{"command":"config","subcommand":"getconfig","token":"d9f2c817-9b8e-4358-9133-995e611b09ab"}' | websocat -n1 ws://localhost:8090
{"command":"config","error":"Errors during specific message validation, please consult the HyperHDR Log","success":false,"tan":0}
Log: WEBSOCKET              : <ERROR> While validating schema against json data of 'JsonRpc@::1':[root].token: no schema definition
```
OK, there is a login subcommand, let's try this:
```
echo '{"command":"authorize","subcommand":"login","token":"d9f2c817-9b8e-4358-9133-995e611b09ab"}' | websocat -n1 ws://localhost:8090
{"command":"authorize-login","success":true,"tan":0}
```

Oh, we are logged in now, lets try to get config:
```
echo '{"command":"config","subcommand":"getconfig"}' | websocat -n1 ws://localhost:8090
{"command":"config","error":"No Authorization","success":false,"tan":0}

echo '{"command":"config","subcommand":"getconfig","token":"d9f2c817-9b8e-4358-9133-995e611b09ab"}' | websocat -n1 ws://localhost:8090
{"command":"config","error":"Errors during specific message validation, please consult the HyperHDR Log","success":false,"tan":0}
Log: WEBSOCKET              : <ERROR> While validating schema against json data of 'JsonRpc@::1':[root].token: no schema definition
```

I don't get it....

wbr

**Comment by awawa-dev** [comment]
> OK, there is a login subcommand, let's try this:
echo '{"command":"authorize","subcommand":"login","token":"d9f2c817-9b8e-4358-9133-995e611b09ab"}' | websocat -n1 ws://localhost:8090
{"command":"authorize-login","success":true,"tan":0}


Token must be longer than 36 chars https://github.com/awawa-dev/HyperHDR/blob/6a6b29dfa970ff1c0b1b4d46192d27deaedab70c/sources/api/JsonAPI.cpp#L1531

otherwise for 36 chars it triggers method that won't unlock admin access only ordinary authorization
https://github.com/awawa-dev/HyperHDR/blob/6a6b29dfa970ff1c0b1b4d46192d27deaedab70c/sources/api/JsonAPI.cpp#L1541

**Comment by Thinner77** [comment]
OK, silly question, how do i get a valid token longer than 36 bytes to unlock admin access?

wbr

**Comment by awawa-dev** [comment]
Did you check 'auth' table for a token for the user? also login with a password returns it https://github.com/awawa-dev/HyperHDR/blob/6a6b29dfa970ff1c0b1b4d46192d27deaedab70c/sources/api/JsonAPI.cpp#L1562

**Comment by Thinner77** [comment]
Hi,

yes there is another entry in auth table
```
        user = Hyperhdr
    password = 4e0d2fa2cf3741d8999b884f5b77dcbe70c1978abbc9ee656fe1046eb08b788fac4db5903453735f77302358fce50ef51bbe5ea43a132c8d6fc713cf1f8d5860
       token = df7d84dabef464722b5253b4187489e82260def11b16e05cafae27958265946d169ad062767754d17f13a4e424b0c35e8cc08849b25b02199d4597581752a90f
        salt = 510c2bc6125aab9a53aa2dd91f655a1903093ff02d335bb565fd32186bad47eb7c73cd597196a78ef6731ad5003ef29271eae5c847de914e96db0c0e283084df
     comment =
portal_token =
          id =
  created_at = 2023-03-23T15:25:33Z
    last_use = 2023-03-29T22:11:56Z
```
looks like token is hashed too, let's try to login with password

```
echo '{"command":"authorize","subcommand":"login","password":"hyperhdr"}' | websocat -n1 ws://localhost:8090

{"command":"authorize-login","info":{"token":"df7d84dabef464722b5253b4187489e82260def11b16e05cafae27958265946d169ad062767754d17f13a4e424b0c35e8cc08849b25b02199d4597581752a90f"},"success":true,"tan":0}
```
hashed token is longer than 36 bytes, maybe we are logged in now?

```
echo '{"command":"config","subcommand":"getconfig"}' | websocat -n1 ws://localhost:8090

{"command":"config","error":"No Authorization","success":false,"tan":0}
```
Maybe the token has to be used either in further requests

```
echo '{"command":"config","subcommand":"getconfig","token":"df7d84dabef464722b5253b4187489e82260def11b16e05cafae27958265946d169ad062767754d17f13a4e424b0c35e8cc08849b25b02199d4597581752a90f"}' | websocat -n1 ws://localhost:8090

{"command":"config","error":"Errors during specific message validation, please consult the HyperHDR Log","success":false,"tan":0}

WEBSOCKET              : <ERROR> While validating schema against json data of 'JsonRpc@::1':[root].token: no schema definition
```
OK key "token" is not defined in schema-config.json, let's try to login with token

```
echo '{"command":"authorize","subcommand":"login","token":"df7d84dabef464722b5253b4187489e82260def11b16e05cafae27958265946d169ad062767754d17f13a4e424b0c35e8cc08849b25b02199d4597581752a90f"}' | websocat -n1 ws://localhost:8090

{"command":"authorize-login","success":true,"tan":0}

echo '{"command":"config","subcommand":"getconfig"}' | websocat -n1 ws://localhost:8090

{"command":"config","error":"No Authorization","success":false,"tan":0}
```

Any more ideas? Code flow looks like this, there are no errors in log (--debug option)

```
void JsonAPI::handleAuthorizeCommand(const QJsonObject& message, const QString& command, int tan)
...
if (subc == "login")
...
			if (token.length() > 36)
			{
				if (API::isUserTokenAuthorized(token))
...

bool API::isUserTokenAuthorized(const QString& userToken)
{
	bool res;
	QMetaObject::invokeMethod(_authManager, "isUserTokenAuthorized", Qt::BlockingQueuedConnection, Q_RETURN_ARG(bool, res), Q_ARG(QString, DEFAULT_CONFIG_USER), Q_ARG(QString, userToken));
	if (res)
	{
		_authorized = true;
		_adminAuthorized = true;
		// Listen for ADMIN ACCESS protected signals
		connect(_authManager, &AuthManager::newPendingTokenRequest, this, &API::onPendingTokenRequest, Qt::UniqueConnection);
	}
	return res;
}

bool AuthManager::isUserAuthorized(const QString& user, const QString& pw)
{
	if (isUserAuthBlocked())
		return false;

	if (!_authTable->isUserAuthorized(user, pw))
	{
		setAuthBlock(true);
		return false;
	}
	return true;
}

bool AuthTable::isUserAuthorized(const QString& user, const QString& pw)
{
	if (userExist(user) && (calcPasswordHashOfUser(user, pw) == getPasswordHashOfUser(user)))
	{
		updateUserUsed(user);
		return true;
	}
	return false;
}
```

wbr

**Comment by awawa-dev** [comment]
![obraz](https://user-images.githubusercontent.com/69086569/228965100-24ac5fc4-ab3b-4d98-b75e-4363940acdc6.png)

https://github.com/vi/websocat/blob/master/moreexamples.md#persistent-client-connection
